onnx/defs/tensor/defs.cc

```diff
@@ -2482,6 +2482,197 @@ ONNX_OPERATOR_SET_SCHEMA(
+            "T1",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .Input(
+            1,
+            "size",
+            "the target output image size (N, C, H, W) for 2D or (N, C, D, H, W) for 3D",
+            "T2",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .Output(
+            0,
+            "grid",
+            "output tensor of shape (N, C, H, W, 2) of 2D sample coordinates or (N, C, D, H, W, 3) of 3D sample coordinates.",
+            "T1",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::Differentiable)
+        .TypeConstraint(
+            "T1",
+            {"tensor(float16)", "tensor(float)", "tensor(double)"},
+            "Constrain grid types to float tensors.")
+        .TypeConstraint("T2", {"tensor(int32)", "tensor(int64)"}, "Constrain indices to integer types")
+        .SetDoc(AffineGrid_ver20_doc)
+        .FunctionBody(R"ONNX(
+        {
+          int_zero = Constant<value_int: int=0>()
+          int_four = Constant<value_int: int=4>()
+
+          constant_align_corners = Constant<value_int: int=@align_corners>()
+          constant_align_corners_equal_zero = Equal (constant_align_corners, int_zero) 
+          
+          size_ndim = Size (size)
+          condition_is_2d = Equal (size_ndim, int_four)
+
+          grid = If (condition_is_2d) <
+              then_branch = g1 () => (float[N, H, W, 2] grid_2d_then) {
+                  minus_one = Constant <value = float {-1.0}>()
+                  zero = Constant <value = float {0.0}>()
+                  one = Constant <value = float {1.0}>()
+                  two = Constant <value = float {2.0}>()
+                  N, C, H, W = Split <num_outputs: int=4>(size)
+                  int_two_1d = Constant<value_ints=[2]>()
+                  int_four_1d = Constant<value_ints=[4]>()
+                  constant_H_W_shape = Slice(size, int_two_1d, int_four_1d) # [N, C, H, W] => [H, W]
+                  zero_H_by_W = ConstantOfShape (constant_H_W_shape)
+                  ones_H_by_W = Add (zero_H_by_W, one)
+
+                  H_float = CastLike (H, zero)
+                  W_float = CastLike (W, zero)
+                  grid_h_0, grid_w_0 = If (constant_align_corners_equal_zero) <
+                      then_branch = h1 () => (float[H] grid_h_then, float[W] grid_w_then) {
+                          step_h = Div (two, H_float)
+                          step_w = Div (two, W_float)
+                          step_h_half = Div (step_h, two)
+                          start_h = Add (minus_one, step_h_half)
+                          grid_h_then = Range (start_h, one, step_h)
+
+                          step_w_half = Div (step_w, two)
+                          start_w = Add (minus_one, step_w_half)
+                          grid_w_then = Range (start_w, one, step_w)
```

Comment:
I am a bit concerned about the use of Range with float types. Couldn't numeric precision affect whether the number of elements generated is off-by-one ? Would it be better to use an integer-based range that is converted into a float range?

Reply:
step_w does not seem to be an integer.

Reply:
Yes, but we know that we want to create a vector of size W. It seems better to use `Range(0, W, 1)` and map each i in this range to a corresponding coordinate-value using arithmetic.

Reply:
Basically, something like
```
   i= Range (zero_int, W, one_int)
   i_step_w = Mul (i, step_w)
   grid_w_then = Add (start_w, i_step_w)
```

Reply:
That may help avoid all the use of `epsilon` in the else-branch. Further, the above fragment is common across the if-then-else, only start_w and step_w depend on the flag.

Reply:
Great suggestion @gramalingam !, I will do the change accordingly.